server/app/services/applicant/ApplicantService.java

```diff
@@ -283,6 +290,181 @@ private CompletionStage<ReadOnlyApplicantProgramService> stageAndUpdateIfValid(
+          Applicant applicant,
+          ImmutableMap<String, String> updateMap,
+          ImmutableSet<Update> updates,
+          Block blockBeforeUpdate,
+          ProgramDefinition programDefinition,
+          boolean addressServiceAreaValidationEnabled) {
+    if (addressServiceAreaValidationEnabled
+        && blockBeforeUpdate.getLeafAddressNodeServiceAreaIds().isPresent()) {
+      Optional<ImmutableList<EsriServiceAreaValidationOption>> maybeOptions =
+          esriServiceAreaValidationConfig.getOptionsByServiceAreaIds(
+              blockBeforeUpdate.getLeafAddressNodeServiceAreaIds().get());
+
+      Optional<ApplicantQuestion> maybeAddressQuestion =
+          blockBeforeUpdate.getAddressQuestionWithCorrectionEnabled();
+      if (maybeAddressQuestion.isPresent() && maybeOptions.isPresent()) {
+        ImmutableList<EsriServiceAreaValidationOption> serviceAreaOptions = maybeOptions.get();
+        ApplicantQuestion addressQuestion = maybeAddressQuestion.get();
+        Boolean hasCorrectedAddress = doesUpdateContainCorrectedAddress(addressQuestion, updateMap);
+
+        if (hasCorrectedAddress) {
+          Path serviceAreaPath = addressQuestion.getContextualizedPath().join(Scalar.SERVICE_AREA);
+          ImmutableList<ServiceAreaInclusion> existingServiceAreaInclusionGroup =
+              getExistingServiceAreaInclusionGroup(serviceAreaPath, updateMap);
+          // filter out the serviceAreaOptions that have already been validated for this address
+          ImmutableList<EsriServiceAreaValidationOption> filteredServiceAreaOptions =
+              serviceAreaOptions.stream()
+                  .filter(
+                      (option) ->
+                          !option.isServiceAreaOptionInInclusionGroup(
+                              existingServiceAreaInclusionGroup))
+                  .collect(ImmutableList.toImmutableList());
+          if (filteredServiceAreaOptions.size() > 0) {
+            AddressLocation addressLocation =
+                getAddressLocationFromUpdates(addressQuestion, updateMap);
+
+            ImmutableList<CompletionStage<ImmutableList<ServiceAreaInclusion>>>
+                serviceAreaInclusionGroupFutures =
+                    filteredServiceAreaOptions.stream()
+                        .map(
+                            (option) -> {
+                              return esriClient
+                                  .getServiceAreaInclusionGroup(option, addressLocation)
+                                  .thenApplyAsync(
+                                      (serviceAreaInclusionGroup) -> serviceAreaInclusionGroup,
+                                      httpExecutionContext.current());
+                            })
+                        .collect(ImmutableList.toImmutableList());
+
+            return CompletableFuture.allOf(
+                    serviceAreaInclusionGroupFutures.toArray(
+                        new CompletableFuture[serviceAreaInclusionGroupFutures.size()]))
```

Comment:
**question (blocking):** is this making a separate ESRI call for each service area? if so, is that necessary? IIRC `EsriClient` checks each service area in a single call

Reply:
It makes a separate esri call for each unique URL among the service area options. Presumably, if there are multiple service area validation endpoints specified, they can contain a different set of boundaries/features.

The new EsriServiceAreaValidationConfig.getOptionsByServiceAreaIds method takes the service area IDs configured for the program address question and returns one EsriServiceAreaValidationOption for each unique URL in the relevant EsriServiceAreaValidationOptions.